pkg/payforput/payments.go

```diff
@@ -0,0 +1,211 @@
+package payforput
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/cashweb/keyserver/pkg/models"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/rs/zerolog/hlog"
+)
+
+// ValidatorFunc is the type of unction which can indicate if a request has
+// the appropriate headers to have been paid.
+type ValidatorFunc func(r *http.Request, secret string) bool
+
+// PaymentEnforcer ensures that the request has been paid for properly by
+// checking for a payment authorization. If there is no payment authorization
+// header, then we send a BIP70 payment request
+type PaymentEnforcer struct {
+	// Location to where we should redirect payments, and has the PaymentHandler
+	// installed
+	PaymentURL string
+	// Validator is a func that validates that a request has a valid proof of
+	// payment
+	Validator ValidatorFunc
+	// Secret is the HMAC secret used for generating and validating tokens
+	// NOTE: This may be swapped out at a later date.
+	Secret string
+}
+
+// New returns a new payment enforcer that can be used for easy BIP70 integration
+// for the keyserver.
+func New(PaymentURL string, Validator ValidatorFunc) *PaymentEnforcer {
+	pe := &PaymentEnforcer{
+		PaymentURL: PaymentURL,
+		Validator:  Validator,
+		Secret:     "",
+	}
+	if pe.Validator == nil {
+		pe.Validator = DefaultValidator
+	}
+	// Generate an ephemeral secret and hold on to it
+	if pe.Secret == "" {
+		pe.Secret = RandString(64)
+	}
+
+	return pe
+}
+
+// DefaultValidator is the default request payment validator
+func DefaultValidator(r *http.Request, secret string) bool {
+	// First attempt to get the code from the querystring
+	token := r.URL.Query().Get("code")
```

Comment:
You could imagine this colliding with some other feature in the future perhaps and adds complexity to signing. What does this provide?

Reply:
This is part of the oauth standard.  It also enables simply following the redirect.